[Doc Feature][Enable slash commands for agents] (#14182)
MicrosoftDocs/msteams-docs: msteams-platform/agents-in-teams/agent-slash-commands.md

                "scopes": ["personal", "team", "groupChat"],
                "supportsTargetedMessages": true
            }
        ]
    }
    ```

<a id="command-list"></a>

- **Provide an explicit command list**: Define a curated set of commands (for example, `/help`, `/create`, `/design`) that appear in the slash menu with a short description. Existing agent commands can be reused, or you can introduce new commands optimized for slash usage.

  After you enable slash commands, declare each command in the manifest, including the command name and a user-facing description. List the specific commands, and not broad categories. Once you declare a command, users can invoke it (for example, <`/create`> or <`/app-name create`>, depending on the client experience).

  You can declare a command list using the new `triggers` property for your agent in one of the following scenarios:

  # [Scenario 1](#tab/sc1)

  **Scenario 1**: Agent with separate @mention and slash command lists

  Use the following example to configure the app manifest for supporting an agent or a bot that offers separate commands for @mention and slash triggers.

    ```json
    {

>         "bots": [

Reviewer: I just read [Create prompt suggestions](https://learn.microsoft.com/en-us/microsoftteams/platform/bots/how-to/conversations/prompt-suggestions) and trying to eyeball the difference between the examples there and this one is tough; it would help if the prose above described the structure of `commandLists` and was prescriptive about what needs to be done, in addition to providing the example (this is true across the board).

This article and "Create prompt suggestions" should link to each other as well, and provide guidance about when to use each one (especially seeing as how "Create prompt suggestions" currently begins with "Prompt suggestions are _commands_ that are presented to the users in the Microsoft Teams chat." (emphasis mine).

Also - can this all be configured in the developer portal too?

Author: @nickwalkmsft There was a link for Action.Submit in the Create prompt suggestion page in the feature article. However, that page lives in the bot module. We'd rather have the content linked from Teams SDK learn docs. That update in Teams SDK is in progress as of now, and once that is published, we'd add the link.

                        "context": ["compose", "commandBox"]
                    },
                    {
                        "id": "draftCommand",
                        "type": "action",
                        "title": "Draft",
                        "description": "Draft a document",
                        "context": ["compose", "commandBox"],
                        "triggers": ["slash"]
                    },
                    {
                        "id": "reviewCommand",
                        "type": "action",
                        "title": "Review",
                        "description": "Review a document",
                        "context": ["compose", "commandBox"],
                        "triggers": ["slash"]
                    }
                ]
            }
        ]
    }
    ```

### Handle slash commands

You can enable the agent to send a private or a public message. You can also enable the agent to update or delete a message that it had previously sent.

#### Send an agent response

Use the following code snippets to enable your agent to respond to a slash command based on [supported scenarios](#supported-scenarios-for-slash-commands):

> [WIP: Add link to Teams SDK docs]

Reviewer: Please do not insert any new links to the SDK docs; we are going to be working on integrating the contents of the SDK docs into the platform docs and retiring the separate SDK docs. Any information in any SDK docs that you would be linking to from here should be represented in this page.

---
title: Enable Slash Commands for Agents
description: Learn how to enable slash commands for your agents and apps
ms.localizationpriority: high
ms.date: 03/11/2026
ms.topic: reference
---

<!-- markdownlint-disable MD036 -->
<!-- markdownlint-disable MD046 -->
<!-- markdownlint-disable MD001 -->
<!-- markdownlint-disable MD024 -->

> # Enable slash commands for agents and apps

Reviewer: Is "and apps" correct/needed here and throughout the article?

Should "slash commands" be "agent slash commands" everywhere, to distinguish? Especially in places like "When a user enters a slash command, it appears in the chat...", which is not true for base slash commands.

Author: @nickwalkmsft Slash commands can be implemented in agents, and bot and message extension apps. For this reason, 'apps' has been added through the article.

> [!NOTE]
>
> Support for slash commands is available in [public developer preview](../resources/dev-preview/developer-preview-intro.md).

> Use slash commands in Teams as a clear, predictable way to interact with agents or apps with simple text commands.

Reviewer: simple -> convenient

This reads as addressing end users; it should address developers who are (considering) implementing slash commands.

Author: Updated.

**Key points**

- [What is a slash command](#what-is-a-slash-command)
- [User experience for slash commands](#user-experience-for-slash-commands)
- **Slash commands developer experience**
  - [Supported scenarios for slash commands](#supported-scenarios-for-slash-commands)
  - [Manifest updates](#update-app-manifest-for-slash-commands)
  - [Handle slash commands](#handle-slash-commands)
  - [Response codes for slash commands](#response-codes-for-slash-commands)
<!--- [Why use slash commands](#why-use-slash-commands)-->

## What is a slash command

> A slash command is a shortcut initiated by typing a forward slash (`/`) in the compose box on Teams, for example, `/contoso`. You can add it to your agent as a command, with optional parameters, to trigger specific actions. Teams shows a menu of available commands from built-in features and any installed agents or apps.

Reviewer: ```suggestion
Slash commands are text-based shortcuts for various actions in Teams, activated from the compose box in chats and meetings. When a user enters a forward slash (`/`) in a compose box, Teams will display a menu of available commands. 

Teams agents that are designed to respond to specific named commands can implement them as slash commands to have them advertised in this menu, alongside the commands offered directly by Teams.
```
This is a rough suggestion, not necessarily final text. EDIT: May need revision based on the ‼️ comment further down.

Also:
- Needs a screenshot of the user experience, so developers understand the experience they're providing.
- See the review comment in the Best Practices section at the bottom. Perhaps this should be "Teams agents that offer command-based functionality _should_ implement it as..."
- The existing user-facing docs page for slash commands is [here](https://support.microsoft.com/en-us/office/use-commands-in-microsoft-teams-88f61508-284d-417f-a53d-9e082164050b) for reference, and it might be worth linking to from here as well. (@luywang are you the one to inform if that potentially needs updates, or if there needs to be a new user-facing article about slash commands for agents?
- I noticed the above page says that slash commands don't work on mobile. Do slash commands _for agents_?

Reviewer 2: Slash for agents will work on mobile. The core ones are still not supported.

> For more information, see [supported scenarios for slash commands](#supported-scenarios-for-slash-commands).

Reviewer: ```suggestion
```

Reviewer: Overall doc design feedback - there are too many links that bounce the reader around to different anchors in the article, these should be used very sparingly.

## User experience for slash commands

> 

Reviewer: ```suggestion
```

> When a user enters a slash command, it appears as a private message visible only to them. The agent can reply privately to a user or, when it's relevant, share the response with the whole group or channel. You can manage who sees responses using the agent logic.

Reviewer: ```suggestion
When a user enters an agent slash command, it appears in the chat as a private message visible only to the sending user and the receiving agent. ((Give brief, one-sentence insight here into the dev experience - what event is fired?))

Your agent's implementation determines the action it takes in response to a slash command and how it replies, including whether that reply is sent privately as a targeted message or as a public message visible to the entire chat. Agent slash command activations are only ever visible to the user who sent them, so consider whether the agent's response should be a private message back to the user. ((See best practices section))
```

Are the private messages _targeted_ messages, or are they something wholly different? If they are targeted, use that vocabulary and maybe link to the targeted messages article.

Re: `((Give brief, one-sentence insight here into the dev experience - what event is fired?))`, this is really key to helping the audience understand what the point of this is (if there is any?) from a developer/implementation point of view. Do these events show up into a Teams SDK-developed agent in a first-class way, or do I need to do my own magic-string parsing of an incoming message, etc.

> :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/agent-slash-commands.png" alt-text="Image shows the response flows for agent slash commands." border="false" lightbox="../assets/images/agents-in-teams/agent-slash-commands/agent-slash-commands.png":::

Reviewer: Replace with clear screenshot of experience and move a little higher up

> [!NOTE]
>
> - Public responses must be used when visibility in the group or channel adds value.

> > - The private message expires after 24 hours.

Reviewer: - "must" carries a lot of technical weight in a developer article and shouldn't be used here.
- What happens when a message expires? Does this apply to all private/targeted messages?

Author: Updated

### Agent response and prompt preview visibility

> Your agents can send a private or public response to a user's query. You can also choose to include prompt preview in agent responses. You can manage the visibility of agent responses to slash commands and prompt preview using the defined response flows:

Reviewer: The "agent response" flow suggestions should be moved to a subsection of the best practices section at the bottom; I'll withhold substantive feedback on that part until that happens. Prompt preview can be left here as a very short subfeature section.

Reviewer 2: IMO this section is pretty important to stand on its own. It talks about the options available to a developer when they choose to reply. They can reply privately or publicly.

> # [Private agent-to-user response](#tab/private)

Reviewer: Remove the tabs UX, it's not called for here.

Author: Tab switcher is used for agent response flow alternatives.

<a id="private-response-1"></a>

This flow keeps slash command results focused between the user and the agent. Use private response flow for drafts, summaries, personal tasks.

:::row:::
    :::column span="2":::
        1. The initial slash command shows up as a private message.
        1. If the agent responds privately, the reply appears only to the initiating user and quotes the user’s prompt. You can include an option for the user to share it publicly if they want.
    :::column-end:::
    :::column span="3":::
        :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/agent-private-response.png" alt-text="Image shows an agent's private response to a user's slash command." border="false" lightbox="../assets/images/agents-in-teams/agent-slash-commands/agent-private-response.png":::
    :::column-end:::
:::row-end:::
:::row:::
    :::column span:::

>         3. If the user chooses to allow, the agent posts a single message with its response along with the prompt preview.

Reviewer: Is it described anywhere how to do this? Is it a built-in interaction?

<a id="public-response-1"></a>

When the response is useful to the wider audience, you can choose to enable your agent to show it publicly.

:::row:::
    :::column span="2":::
        1. When a user runs a slash command, it appears right away as a private message.
        1. If the agent is configured for relevant public replies, the response is posted to all members in the group or channel. See [best practices](#best-practices) for more guidelines.
        1. The agent responds with the prompt preview.
    :::column-end:::
    :::column span="3":::
        :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/agent-public-response.png" alt-text="Image shows agent's public response." border="false" lightbox="../assets/images/agents-in-teams/agent-slash-commands/agent-public-response.png":::
    :::column-end:::
:::row-end:::
:::row:::
    :::column span:::
        This flow is useful for sharing status updates, confirmations, or other information meant for everyone in group or channel.
    :::column-end:::
:::row-end:::

Next, enable [public agent-to-user responses](#send-an-agent-response).

# [Prompt preview](#tab/preview)

> When an agent replies to a user, prompt preview displays the user’s initial slash command query in a single self-contained message:

Reviewer: What *is* prompt preview? Is it an implementation choice, internal app configuration choice, agent setting, user option, global client setting?

Reviewer 2: Its an optional UX treatment to show the user's prompt along with the bot response. The bot can optionally leave it out.

Reviewer: > The bot can optionally leave it out.

My point here is that the scope at which this choice is available isn't clear, like if this is a setting that's applied to the whole agent. That can easily be fixed by rephrasing to something like "With each response, you can opt to include a _prompt preview_: the user's initial slash command query, shown in an inline message block at the top of the response message."

Also, this definition should be lifted up out of this tab into the text that's just below the heading of this section.

> - **Private agent-to-user response**: The agent replies privately to the user's targeted message, so only the intended user can see the reply.

Reviewer: Not clear if this referring to an overall private response, or that the prompt preview part can be made private within an overall public response.

  :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/private-prompt-preview.png" alt-text="Image shows the prompt preview for private agent-to-user response.":::

- **Public agent-to-user response**: The agent sends a public reply to the user's message that is visible to everyone in the chat.

  :::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/public-prompt-preview.png" alt-text="Image shows the prompt preview for public agent-to-user response.":::

> The prompt preview is always included at the top of the agent response. A single agent response can include multiple prompt previews. However, it appears doesn't appear for normal messages.

Reviewer: ```suggestion
When included, the prompt preview is always placed at the top of the agent response. A single agent response can include multiple prompt previews. However, it appears doesn't appear for normal messages.
```

- When/why would it include multiple prompt previews?
- What is a "normal message" and why wouldn't it appear there?

> [!NOTE]

> > Including the user’s prompt in the agent response is optional. Use prompt preview to maintain context in ongoing conversations.

Reviewer: Remove the note, include this assertion above and in best practices.

Author: @nickwalkmsft Updated

> Next, enable [prompt preview](#send-an-agent-response).

Reviewer: ```suggestion
```

Author: Updated as suggested.

---

<!--
## Why use slash commands

> Slash commands make agents easier to use in shared chats. They let people run common actions with short commands and quickly see what’s available by typing `/`. Because the commands are clear and consistent, they reduce confusion and work within defined permissions. Your agent can also respond in a private one-turn conversation, which helps keep busy conversations less noisy.

Reviewer: Include in best practices

Author: @nickwalkmsft This part has been commented out in one of the previous reviews.

-->

> ## Slash commands developer experience

Reviewer: ```suggestion
## Implement slash commands
```

Author: Updated

Enabling slash commands typically involves the following:

- [Select one or more of the supported scenarios for enabling slash commands](#supported-scenarios-for-slash-commands): You can enable slash commands for your agents, bots, and message extension apps.
- [Update app manifest and declare supported commands](#update-app-manifest-for-slash-commands): You can opt for slash commands through the Teams app manifest. You can optionally declare commands in the app manifest to present users with a curated set of slash commands for triggering actions via <`/app-name`>.
- [Handle slash commands in your agent](#handle-slash-commands): Use Teams SDK or REST APIs to handle the user's request and to send the agent response.

> ### Supported scenarios for slash commands

Reviewer: ‼️  This section is very hard to follow and has big ramifications for the rest of the article:

- Is this design guidance or description of technical limitations?
- Are "user to agent interaction" and "app-defined slash commands" mutually exclusive? It seems like they are, in which case these represent two very distinct "modes" by which an agent can offer slash commands (I wouldn't even call the former mode "slash commands", it's more like a general private message feature) that are so functionally different that they should be spelled out near the top of the article, and the overall article designed around them. Are they mutually exclusive with "Message extension actions as slash commands"?

I'm withholding feedback on much of the stuff below related to this because the way it works is critical to how it's worded.

EDIT: After reading the targeted messages doc: assuming I'm understanding this all correctly, I think everything about "user to agent interaction" should be moved to that article, because it's effectively just a user sending a targeted message to an agent. The concept of advertised commands, via slash _or_ @ mention, should be the focus of this article, and they can refer to each other as appropriate.

Reviewer 2: User to agent and app defined slash commands are not mutually exclusive. We have three slash command options:
1. user to agent commands with /<agent-name> for natural language prompts eg /copilot 
2. app defined slash commands with /<command-name> for a defined command eg /subscribe 
3. ME of action types

Reviewer 2: We can take a crack at improving the readability of this section. Also the way I consider targeted messages and slash commands is that TMs can exist outside of slash. 
Slash is user driven but agent defined. It uses TMs under the hood but that does not matter to developers. 

The developer story on this article is opt into slash commands with the manifest changes listed - in 3 ways (all or any). Handle the incoming slash command. Respond to the user privately (TM) or publicly. Here are the best practises in replying back to the user

Reviewer: Let me back up and approach this from a different angle:

After reading through the text about user experience, scenarios and workflows, it's surprising to then read and grok the implementation part to see that there are really only two things going on here from a developer perspective:

1. `supportsTargetedMessages` (along with the `IsTargeted` property on received messages), which enables users to send freeform targeted messages to the agent. It's not clear why this is in this article (as opposed to the targeted messages doc) and described as an aspect of slash commands: while it's true that this capability is activated on the user side with a `/`, and it's certainly true that developers should understand the experience their app will offer, it doesn't have anything to do with named commands and is just about enabling users to send targeted messages to an agent by invoking its name with a `/`.
1. Listing commands to advertise them to users via autocomplete, and in the case of slash commands (as opposed to at-mention commands), enabling the client behavior of sending a targeted message to the agent when they are activated.

Everything beyond that is guidance - it's good guidance and clearly critical for this feature, but it's integrated into the article in a way that makes it difficult to reach an understanding of exactly what the platform capabilities are vs. the directed guidance for how to use it.

It's also surprising to reach the realization that _commands_ aren't actually a first-class thing in the activity protocol or programming model, it's all just messages, and it's on the developer to implement the concept of named commands on top of messaging. That's not necessarily a problem, but much of the article reads like commanding is a first-class feature all the way through.

You can enable slash commands for your agents in the following ways:

- **User-to-agent interaction**: Slash commands can initiate one-turn interaction with agent. In this model, the user enters a command (and optional text) in the compose box, and the response is delivered privately, making it ideal for drafting, lookups, and personal productivity tasks. For example:

  - `/contoso incident summarize the last 24 hours and suggest next steps`
  - `/contoso create-task fix login issue for mobile users`

  To [support natural-language prompts](#command-list), you must explicitly opt in so agents that want to support only fixed commands can do so without enabling free-form prompting.

> - **App-defined slash commands**: Similar to using @mentions in the compose box, agents can publish a curated set of [slash commands](#command-list). Users can discover and run common actions without leaving the compose box. They can explicitly declare the commands your agent supports, and Teams shows them in the slash command picker when a user types `/`. For example:

Reviewer: Seems like this should be the leading scenario, and the freeform text one should be the secondary. Also shouldn't the examples lead with an agent name?

Are explicitly-supported and advertised commands via @mention a new feature too?

  - `/help`
  - `/settings`

> - **Message extension actions as slash commands**: You can [surface action-type message extensions as slash commands](#me). When a user selects the command from the `/` menu, Teams opens the associated modal (task module) or dialog so the user can complete the action with guided inputs, validation, and a consistent UI flow. For example, `/contoso create task` (opens a task creation dialog).

Reviewer: This is cool and should be mentioned as a possibility in the lead-in at the top.

  > [!NOTE]

>   > Search type message extensions aren’t exposed as slash commands. Users can access them through the message extension flyout instead.

Reviewer: Can this link to something? Not clear what "search type message extensions" or "message extension flyout" are.

  You can declare a command list using the new `triggers` property for your agent in one of the following scenarios:

  # [Scenario 1](#tab/sc1)

  **Scenario 1**: Agent with separate @mention and slash command lists

  Use the following example to configure the app manifest for supporting an agent or a bot that offers separate commands for @mention and slash triggers.

    ```json
    {
        "bots": [
            {
                "botId": "{{BOT_ID}}",
                "scopes": ["personal", "team", "groupChat"],
                "supportsTargetedMessages": true,
                "commandLists": [
                    {
                        "scopes": ["personal", "team", "groupChat"],
                        "triggers": ["mention"],
                        "commands": [
                            { "title": "Summarize", "description": "Summarize a document" },
                            { "title": "Draft", "description": "Draft a document" }
                        ]
                    },
                    {

>                         "scopes": ["team", "gro54upChat"],

Reviewer 3: typo: groupChat

Author: Updated

                ]
            }
        ]
    }
    ```

<a id="me"></a>

- **Enable message extension actions as slash commands**: Use the following example to configure the app manifest for supporting message extension actions as slash commands using the new `triggers` property.

  > [!NOTE]
  > You can configure the app manifest to enable slash commands for both the bot and message extension by combining the manifest examples from scenarios 1 or 2 with scenario 3.

    ```json
    {
        "composeExtensions": [
            {
                "botId": "{{BOT_ID}}",
                "commands": [
                    {
                        "id": "summarizeCommand",
                        "type": "action",
                        "title": "Summarize",
                        "description": "Summarize a document",
                        "context": ["compose", "commandBox"]
                    },
                    {
                        "id": "draftCommand",
                        "type": "action",
                        "title": "Draft",
                        "description": "Draft a document",
                        "context": ["compose", "commandBox"],

>                         "triggers": ["slash"]

Reviewer 3: Clarify if this configuration means it's accessible as *both* a slash command and ME action, or only via slash.

                    },
                    {
                        "id": "reviewCommand",
                        "type": "action",
                        "title": "Review",
                        "description": "Review a document",
                        "context": ["compose", "commandBox"],
                        "triggers": ["slash"]
                    }
                ]
            }
        ]
    }
    ```

### Handle slash commands

> You can enable the agent to send a private or a public message. You can also enable the agent to update or delete a message that it had previously sent.

Reviewer: "Enable" (and "configure", both throughout) imply configuration and settings, and aren't good verbs where writing code is involved, unless the code involves manipulating configuration and settings.

This all looks to be general documentation and/or guidance about sending responses in general; where is the documentation specifically about handling commands? If handling and responding to commands mostly looks and feels like handling and responding to regular messages, it should be framed that way, and emphasize the difference

    async def handle_message(ctx):
      if getattr(ctx.activity.recipient, "is_targeted", False):
        await ctx.send(MessageActivityInput("Reactive TM").with_recipient(Account(id="<userMRI>", name="<user Name>", role=Role.USER), is_targeted=True))
  ```

  # [HTTP](#tab/api1)

  [WIP: Add code snippet]

# [Public response by the agent](#tab/public)

You can enable the agent to send for the [public response scenario](#agent-response-and-prompt-preview-visibility) in a group or a channel if:

- The message requires collaboration from all members.
- The broader visibility adds value.

  # [C#](#tab/dotnet)

    ```csharp
  
      teams.OnMessage(async (context, cancellationToken) => {
        await context.Send(new MessageActivity("Normal msg"), cancellationToken);
      });
    ```
  
  # [TypeScript](#tab/ts)

    ```typescript

    app.on('message', async ({ send, activity }) => {
      send(new MessageActivity('Normal msg'))  
    });
    ```

  # [Python](#tab/Py)

    ```python
  
      @app.on_message
      async def handle_message(ctx):
        await ctx.send(MessageActivityInput("Normal msg"))
    ```
  
  # [HTTP](#tab/api)

    [WIP: Add code snippet]

> # [Prompt preview](#tab/preview)

Reviewer: TODO, this appears to be a placecholder

  async def handle_message_delete(ctx):
    if getattr(ctx.activity.recipient, "is_targeted", False):
      # Business logic when message hard delete flow is for TM
  ```

# [HTTP](#tab/api1)

  [WIP: Add code snippet]

---

## Response codes for slash commands

Ensure to handle these errors appropriately in your agent. The following table lists error codes, error descriptions, and developer actions for Teams SDK:

| Status code | Error code | Description | Developer action |
| --- | --- | --- | --- | --- |
| 400 | `INVALID_TARGETED_MESSAGE_ID` | The message ID used for the prompt preview is invalid. | Ensure that the message ID for the targeted message is correct. |
| 404 | `TARGETED_MESSAGE_EXPIRED_OR_DELETED` | The message ID associated with the prompt preview in the agent response could not be found in the conversation. The message is unavailable as it was deleted or auto removed after 24 hours. | Ensure the agent either sends a new targeted message or waits for user input, as per business logic. |

You can also see more information on [error codes for targeted messages](targeted-messages.md#handle-errors).

## Code sample

[WIP: Add details and link to the code sample.]

> ## Best practices

Reviewer: ```suggestion
## Agent design guidance and best practices
```

Seems like best practice zero is "If your agent has command-based functionality, you should implement it as slash commands to make it a first-class thing recognized by the platform"? If that's the case, it should be stated near the top of the article, and again here.

Also, as discussed during the meeting - consideration around whether the agent should reply privately or not.

---
title: Enable Slash Commands for Agents
description: Learn how to enable slash commands for your agents and apps
ms.localizationpriority: high
ms.date: 03/11/2026
ms.topic: reference
---

<!-- markdownlint-disable MD036 -->
<!-- markdownlint-disable MD046 -->
<!-- markdownlint-disable MD001 -->
<!-- markdownlint-disable MD024 -->

# Enable slash commands for agents and apps

> [!NOTE]
>
> Support for slash commands is available in [public developer preview](../resources/dev-preview/developer-preview-intro.md).

> Use slash commands in Teams as a clear, predictable way to interact with agents or apps with convenient text commands.

Reviewer: ```suggestion
_Agent slash commands_ in Teams help users discover clear, predictable ways to interact with agents or apps using convenient text commands.
```

**Key points**

- [What is a slash command](#what-is-a-slash-command)
- **Slash commands developer experience**
  - [Supported scenarios for slash commands](#supported-scenarios-for-slash-commands)
  - [Implement slash commands](#implement-slash-commands)
  - [Manifest updates](#update-app-manifest-for-slash-commands)

## What is a slash command

Slash commands are text-based shortcuts in Teams that let users perform specific actions from the compose box in chats and meetings. When a user types a forward slash (`/`) in the compose box, Teams shows a menu of [available commands](https://support.microsoft.com/en-us/office/use-commands-in-microsoft-teams-88f61508-284d-417f-a53d-9e082164050b).

> Teams agents designed to respond to specific named commands can be implemented as slash commands with optional parameters. Slash commands for agents are supported for mobile clients across platforms.

Reviewer: ```suggestion
Slash commands are text-based shortcuts in Teams that let users perform specific actions from the compose box in chats and meetings. When a user types a forward slash (`/`) in the compose box, Teams shows an autocomplete menu of available commands. The list includes a set of built-in [commands for common tasks](https://support.microsoft.com/office/use-commands-in-microsoft-teams-88f61508-284d-417f-a53d-9e082164050b), as well as the names of agents that are eligible to receive targeted messages. Users use these agent-named commands to send targeted messages to agents (TODO article link).

You can configure your app or agent to add additional items to the list to help users discover command-style behavior that it offers. Three types of commands can be added:

- **Conversational slash commands**: when activated, these commands populate the compose box with a targeted (private) message (TODO link) to the agent that includes the name of the command. Users can modify or append to the text before sending the message. Conversational slash commands are useful for surfacing functionality that users may want to use privately in group conversations.
- **Conversational @mention commands**: these function just like conversational slash commands, but are listed in an autocomplete box that appears when @mentioning the agent. Activating one populates the compose box with a public message that @mentions the agent and includes the name of the command.
- **Message extension slash commands**: apps that implement action-type message extensions can surface them as discoverable slash commands. When activated, message extension slash commands immediately open the associated task module or dialog and do not populate the compose box or send a message. (TODO I think these can be activated from the command box too?). Search-type message extensions cannot be exposed as commands.

Slash command names do not need to be globally unique. Each command in the autocomplete box is displayed with the name and icon of the agent that provides them, enabling users to differentiate between them.

Unlike the built-in slash commands, agent commands are supported across mobile platforms.
```

Author: @nickwalkmsft The scenarios for slash commands are already covered under the section Supported scenarios for slash commands. @mentions are listed under best practices as it isn't directly relevant to slash commands.

:::image type="content" source="../assets/images/agents-in-teams/agent-slash-commands/slash-command-compose-box.png" alt-text="Image shows the response flows for agent slash commands." border="false" lightbox="../assets/images/agents-in-teams/agent-slash-commands/slash-command-compose-box.png":::

- Slash commands are **visible only to the user** who triggers them.
- Your **agent's implementation** determines both the action and response visibility.
- Agent responses can be **[targeted (private) messages](targeted-messages.md) or public messages** shared with the group/channel.
- Always assess if a **private response is more appropriate** before replying publicly. Use public responses when visibility in the group or channel adds value.

When an agent receives a message, it is handled through the `message` event. The agent can determine whether the message is a targeted message by checking `Recipient.IsTargeted`. If true, it is a targeted message. Slash commands are delivered as `MessageActivity` events within the `OnMessage` handler with `Recipient.IsTargeted = true`.

See [best practices](#best-practices).

> [!NOTE]
>

> > The private message is removed after 24 hours. The user can't view or respond to the private message after that.

Reviewer: ```suggestion
When a user activates a conversational slash or @mention command to populate the compose box and then sends the message, the result is a standard text message to the agent (either targeted or public). The agent's implementation is responsible for interpreting the text as a named command and handling it appropriately. Message extension slash commands activate the extension as normal.

Always consider whether a response should be private or public given the context. See [best practices](#best-practices).

```

Reviewer: I kind of punted on the last bit here, but the message should be "there is a lot to consider here to avoid frustrating or surprising users, make sure to read the best practices and design guidance."

## Implement slash commands

Enabling slash commands typically involves the following:

- [Select one or more of the supported scenarios for enabling slash commands](#supported-scenarios-for-slash-commands): You can enable slash commands for your agents, bots, and message extension apps.
- [Update app manifest and declare supported commands](#update-app-manifest-for-slash-commands): You can opt for slash commands through the Teams app manifest. You can optionally declare commands in the app manifest to present users with a curated set of slash commands for triggering actions via <`/app-name`>.

> ### Supported scenarios for slash commands

Reviewer: This section has been replaced by the edits above and can be removed

                        "title": "Review",
                        "description": "Review a document",
                        "context": ["compose", "commandBox"],
                        "triggers": ["slash"]
                    }
                ]
            }
        ]
    }
    ```

When you implement slash commands, you should also determine how the agent or app responds to the user's query. The agent can respond privately to the user or to the group or channel, as needed. For more information, see [agent response scenarios for slash commands](targeted-messages.md#agent-response-for-slash-commands).

## Best practices

- Keep slash commands short and action-oriented.
- Use slash commands for private or quick actions, and @mentions for richer conversations.
- Avoid turning long prompt suggestions into slash commands. Keep them lightweight.
- Provide clear command descriptions to improve discoverability.
- Add aliases for high-traffic commands where appropriate.

For more information on best practices for determining agent responses for slash commands, see [design guidelines for agent responses](targeted-messages.md#best-practices-for-agent-responses).

> ## Code sample

Reviewer: Article should ideally get an inline example scenario as well, including implementation that shows a named command being interpreted.

[WIP: Add details and link to the code sample.]

## Next step

> [!div class="nextstepaction"]
> [Agent responses for slash commands](targeted-messages.md#agent-response-for-slash-commands)

> <!--

Reviewer: Editor note: article ends here

(Please remove comments from the PR and temporarily store somewhere else when they're this extensive)
